Instanceof pattern variables

There is a basic test [InstanceOfPatternVariable.java](https://github.com/eisop/checker-framework/blob/064d3279bfa8eeb3cf2c129152d5e75be2d840b1/checker/tests/nullness/java17/InstanceOfPatternVariable.java#L11) that ensures that instanceof pattern variables don't crash.

This should be extended with code like this:

````
        if (x instanceof @Nullable String n) {
            n.hashCode();
        }
        if (x instanceof @NonNull Number nn) {
            nn.hashCode();
        }
````

Compiling this with the Nullness Checker gives no warnings - `n` correctly is non-null.

If one removes the uses of `n` and `nn`, the Nullness Checker raises the following warnings:

````
checker/tests/nullness/java17/InstanceOfPatternVariable.java:18: error: [instanceof.nullable] instanceof is only true for a non-null expression
        if (x instanceof @Nullable String) {}
              ^
checker/tests/nullness/java17/InstanceOfPatternVariable.java:19: warning: [instanceof.nonnull.redundant] redundant @NonNull annotation on instanceof
        if (x instanceof @NonNull Number) {}
              ^
1 error
1 warning
````

I would expect the same warnings/errors when using instanceof pattern variables.

--- a/checker/src/main/java/org/checkerframework/checker/nullness/NullnessNoInitVisitor.java
+++ b/checker/src/main/java/org/checkerframework/checker/nullness/NullnessNoInitVisitor.java
@@ -6,7 +6,6 @@ import com.sun.source.tree.ArrayAccessTree;
 import com.sun.source.tree.ArrayTypeTree;
 import com.sun.source.tree.AssertTree;
 import com.sun.source.tree.BinaryTree;
-import com.sun.source.tree.BindingPatternTree;
 import com.sun.source.tree.CaseTree;
 import com.sun.source.tree.CatchTree;
 import com.sun.source.tree.ClassTree;
@@ -23,11 +22,9 @@ import com.sun.source.tree.LiteralTree;
 import com.sun.source.tree.MemberSelectTree;
 import com.sun.source.tree.MethodInvocationTree;
 import com.sun.source.tree.MethodTree;
-import com.sun.source.tree.ModifiersTree;
 import com.sun.source.tree.NewArrayTree;
 import com.sun.source.tree.NewClassTree;
 import com.sun.source.tree.ParameterizedTypeTree;
-import com.sun.source.tree.PatternTree;
 import com.sun.source.tree.SwitchTree;
 import com.sun.source.tree.SynchronizedTree;
 import com.sun.source.tree.ThrowTree;
@@ -58,9 +55,11 @@ import org.checkerframework.javacutil.ElementUtils;
 import org.checkerframework.javacutil.TreePathUtil;
 import org.checkerframework.javacutil.TreeUtils;
 import org.checkerframework.javacutil.TreeUtilsAfterJava11;
+import org.checkerframework.javacutil.TreeUtilsAfterJava11.BindingPatternUtils;
 import org.checkerframework.javacutil.TreeUtilsAfterJava11.SwitchExpressionUtils;
 import org.checkerframework.javacutil.TypesUtils;
 
+import java.util.ArrayList;
 import java.util.List;
 
 import javax.annotation.processing.ProcessingEnvironment;
@@ -475,35 +474,32 @@ public class NullnessNoInitVisitor extends BaseTypeVisitor<NullnessNoInitAnnotat
             return null;
         }
 
-        PatternTree pattern = tree.getPattern();
+        List<AnnotationMirror> annotations = new ArrayList<>();
 
-        if (pattern != null && pattern.getKind() == Tree.Kind.BINDING_PATTERN) {
-            BindingPatternTree bindingPattern = (BindingPatternTree) pattern;
-            VariableTree variable = bindingPattern.getVariable();
-            ModifiersTree modifiers = variable.getModifiers();
+        if (refTypeTree.getKind() == Tree.Kind.ANNOTATED_TYPE) {
+            annotations.addAll(TreeUtils.annotationsFromTree((AnnotatedTypeTree) refTypeTree));
+        }
 
-            List<? extends AnnotationTree> annotationTrees = modifiers.getAnnotations();
-
-            List<? extends AnnotationMirror> annotations =
-                    TreeUtils.annotationsFromTypeAnnotationTrees(annotationTrees);
-
-            if (AnnotationUtils.containsSame(annotations, NULLABLE)) {
-                checker.reportError(tree, "instanceof.nullable");
-            }
-            if (AnnotationUtils.containsSame(annotations, NONNULL)) {
-                checker.reportWarning(tree, "instanceof.nonnull.redundant");
-            }
-
-        } else if (refTypeTree.getKind() == Tree.Kind.ANNOTATED_TYPE) {
-            List<? extends AnnotationMirror> annotations =
-                    TreeUtils.annotationsFromTree((AnnotatedTypeTree) refTypeTree);
-            if (AnnotationUtils.containsSame(annotations, NULLABLE)) {
-                checker.reportError(tree, "instanceof.nullable");
-            }
-            if (AnnotationUtils.containsSame(annotations, NONNULL)) {
-                checker.reportWarning(tree, "instanceof.nonnull.redundant");
+        Tree patternTree = TreeUtilsAfterJava11.InstanceOfUtils.getPattern(tree);
+        if (patternTree != null) {
+            if (TreeUtils.isBindingPatternTree(patternTree)) {
+                VariableTree variableTree = BindingPatternUtils.getVariable(patternTree);
+                if (variableTree.getModifiers() != null) {
+                    List<? extends AnnotationTree> annotationTree =
+                            variableTree.getModifiers().getAnnotations();
+                    annotations.addAll(
+                            TreeUtils.annotationsFromTypeAnnotationTrees(annotationTree));
+                }
             }
         }
+
+        if (AnnotationUtils.containsSame(annotations, NULLABLE)) {
+            checker.reportError(tree, "instanceof.nullable");
+        }
+        if (AnnotationUtils.containsSame(annotations, NONNULL)) {
+            checker.reportWarning(tree, "instanceof.nonnull.redundant");
+        }
+
         // Don't call super because it will issue an incorrect instanceof.unsafe warning.
         return null;
     }

--- a/checker/src/main/java/org/checkerframework/checker/nullness/NullnessNoInitVisitor.java
+++ b/checker/src/main/java/org/checkerframework/checker/nullness/NullnessNoInitVisitor.java
@@ -6,6 +6,7 @@ import com.sun.source.tree.ArrayAccessTree;
 import com.sun.source.tree.ArrayTypeTree;
 import com.sun.source.tree.AssertTree;
 import com.sun.source.tree.BinaryTree;
+import com.sun.source.tree.BindingPatternTree;
 import com.sun.source.tree.CaseTree;
 import com.sun.source.tree.CatchTree;
 import com.sun.source.tree.ClassTree;
@@ -22,9 +23,11 @@ import com.sun.source.tree.LiteralTree;
 import com.sun.source.tree.MemberSelectTree;
 import com.sun.source.tree.MethodInvocationTree;
 import com.sun.source.tree.MethodTree;
+import com.sun.source.tree.ModifiersTree;
 import com.sun.source.tree.NewArrayTree;
 import com.sun.source.tree.NewClassTree;
 import com.sun.source.tree.ParameterizedTypeTree;
+import com.sun.source.tree.PatternTree;
 import com.sun.source.tree.SwitchTree;
 import com.sun.source.tree.SynchronizedTree;
 import com.sun.source.tree.ThrowTree;
@@ -471,7 +474,27 @@ public class NullnessNoInitVisitor extends BaseTypeVisitor<NullnessNoInitAnnotat
             // Handle them properly.
             return null;
         }
-        if (refTypeTree.getKind() == Tree.Kind.ANNOTATED_TYPE) {
+
+        PatternTree pattern = tree.getPattern();
+
+        if (pattern != null && pattern.getKind() == Tree.Kind.BINDING_PATTERN) {
+            BindingPatternTree bindingPattern = (BindingPatternTree) pattern;
+            VariableTree variable = bindingPattern.getVariable();
+            ModifiersTree modifiers = variable.getModifiers();
+
+            List<? extends AnnotationTree> annotationTrees = modifiers.getAnnotations();
+
+            List<? extends AnnotationMirror> annotations =
+                    TreeUtils.annotationsFromTypeAnnotationTrees(annotationTrees);
+
+            if (AnnotationUtils.containsSame(annotations, NULLABLE)) {
+                checker.reportError(tree, "instanceof.nullable");
+            }
+            if (AnnotationUtils.containsSame(annotations, NONNULL)) {
+                checker.reportWarning(tree, "instanceof.nonnull.redundant");
+            }
+
+        } else if (refTypeTree.getKind() == Tree.Kind.ANNOTATED_TYPE) {
             List<? extends AnnotationMirror> annotations =
                     TreeUtils.annotationsFromTree((AnnotatedTypeTree) refTypeTree);
             if (AnnotationUtils.containsSame(annotations, NULLABLE)) {

--- a/checker/src/main/java/org/checkerframework/checker/nullness/NullnessNoInitVisitor.java
+++ b/checker/src/main/java/org/checkerframework/checker/nullness/NullnessNoInitVisitor.java
@@ -59,7 +59,6 @@ import org.checkerframework.javacutil.TreeUtilsAfterJava11.BindingPatternUtils;
 import org.checkerframework.javacutil.TreeUtilsAfterJava11.SwitchExpressionUtils;
 import org.checkerframework.javacutil.TypesUtils;
 
-import java.util.ArrayList;
 import java.util.List;
 
 import javax.annotation.processing.ProcessingEnvironment;
@@ -474,30 +473,28 @@ public class NullnessNoInitVisitor extends BaseTypeVisitor<NullnessNoInitAnnotat
             return null;
         }
 
-        List<AnnotationMirror> annotations = new ArrayList<>();
-
+        List<? extends AnnotationMirror> annotations = null;
         if (refTypeTree.getKind() == Tree.Kind.ANNOTATED_TYPE) {
-            annotations.addAll(TreeUtils.annotationsFromTree((AnnotatedTypeTree) refTypeTree));
-        }
-
-        Tree patternTree = TreeUtilsAfterJava11.InstanceOfUtils.getPattern(tree);
-        if (patternTree != null) {
-            if (TreeUtils.isBindingPatternTree(patternTree)) {
+            annotations = TreeUtils.annotationsFromTree((AnnotatedTypeTree) refTypeTree);
+        } else {
+            Tree patternTree = TreeUtilsAfterJava11.InstanceOfUtils.getPattern(tree);
+            if (patternTree != null && TreeUtils.isBindingPatternTree(patternTree)) {
                 VariableTree variableTree = BindingPatternUtils.getVariable(patternTree);
                 if (variableTree.getModifiers() != null) {
                     List<? extends AnnotationTree> annotationTree =
                             variableTree.getModifiers().getAnnotations();
-                    annotations.addAll(
-                            TreeUtils.annotationsFromTypeAnnotationTrees(annotationTree));
+                    annotations = TreeUtils.annotationsFromTypeAnnotationTrees(annotationTree);
                 }
             }
         }
 
-        if (AnnotationUtils.containsSame(annotations, NULLABLE)) {
-            checker.reportError(tree, "instanceof.nullable");
-        }
-        if (AnnotationUtils.containsSame(annotations, NONNULL)) {
-            checker.reportWarning(tree, "instanceof.nonnull.redundant");
+        if (annotations != null) {
+            if (AnnotationUtils.containsSame(annotations, NULLABLE)) {
+                checker.reportError(tree, "instanceof.nullable");
+            }
+            if (AnnotationUtils.containsSame(annotations, NONNULL)) {
+                checker.reportWarning(tree, "instanceof.nonnull.redundant");
+            }
         }
 
         // Don't call super because it will issue an incorrect instanceof.unsafe warning.

--- a/checker/tests/nullness/java17/NullnessInstanceOf.java
+++ b/checker/tests/nullness/java17/NullnessInstanceOf.java
@@ -1,0 +1,47 @@
+// @below-java14-jdk-skip-test
+import org.checkerframework.checker.nullness.qual.NonNull;
+import org.checkerframework.checker.nullness.qual.Nullable;
+
+public class NullnessInstanceOf {
+
+    public void testClassicInstanceOfNullable(Object x) {
+        // :: error: (instanceof.nullable)
+        if (x instanceof @Nullable String) {
+            System.out.println("Nullable String instanceof check.");
+        }
+    }
+
+    public void testClassicInstanceOfNonNull(Object x) {
+        // :: warning: (instanceof.nonnull.redundant)
+        if (x instanceof @NonNull Number) {
+            System.out.println("NonNull Number instanceof check.");
+        }
+    }
+
+    public void testPatternVariableNullable(Object x) {
+        // :: error: (instanceof.nullable)
+        if (x instanceof @Nullable String n) {
+            System.out.println("Length of String: " + n.length());
+        }
+    }
+
+    public void testPatternVariableNonNull(Object x) {
+        // :: warning: (instanceof.nonnull.redundant)
+        if (x instanceof @NonNull Number nn) {
+            System.out.println("Number's hashCode: " + nn.hashCode());
+        }
+    }
+
+    public void testUnannotatedPatternVariable(Object x) {
+        if (x instanceof String unannotatedString) {
+            System.out.println("Unannotated String length: " + unannotatedString.length());
+        }
+    }
+
+    public void testUnusedPatternVariable(Object x) {
+        // :: error: (instanceof.nullable)
+        if (x instanceof @Nullable String unusedString) {}
+        // :: warning: (instanceof.nonnull.redundant)
+        if (x instanceof @NonNull Number unusedNumber) {}
+    }
+}

--- a/checker/tests/nullness/java17/NullnessInstanceOf.java
+++ b/checker/tests/nullness/java17/NullnessInstanceOf.java
@@ -32,6 +32,12 @@ public class NullnessInstanceOf {
         }
     }
 
+    public void testUnannotatedClassic(Object x) {
+        if (x instanceof String) {
+            System.out.println("Unannotated String instanceof check.");
+        }
+    }
+
     public void testUnannotatedPatternVariable(Object x) {
         if (x instanceof String unannotatedString) {
             System.out.println("Unannotated String length: " + unannotatedString.length());

--- a/docs/CHANGELOG.md
+++ b/docs/CHANGELOG.md
@@ -10,7 +10,7 @@ and a redundant warning if they are annotated with `@NonNull`.
 
 **Closed issues:**
 
-eisop#1003, eisop#1033, eisop#1058
+eisop#1003, eisop#1033, eisop#1058.
 
 
 Version 3.42.0-eisop5 (December 20, 2024)

--- a/docs/CHANGELOG.md
+++ b/docs/CHANGELOG.md
@@ -3,11 +3,14 @@ Version 3.42.0-eisop6 (January ??, 2025)
 
 **User-visible changes:**
 
+The Nullness Checker now reports an error if any instanceof pattern variables are annotated with `@Nullable`
+and a redundant warning if they are annotated with `@NonNull`.
+
 **Implementation details:**
 
 **Closed issues:**
 
-eisop#1003, eisop#1033.
+eisop#1003, eisop#1033, eisop#1058
 
 
 Version 3.42.0-eisop5 (December 20, 2024)

